leo/cli/commands/deploy.rs

```diff
@@ -51,55 +53,62 @@ impl Command for Deploy {
-        //     // Set deploy arguments
-        //     let deploy = SnarkOSDeploy::try_parse_from([
-        //         "snarkos",
-        //         "--private-key",
-        //         private_key.as_ref().unwrap(),
-        //         "--query",
-        //         self.endpoint.as_str(),
-        //         "--priority-fee",
-        //         self.priority_fee.as_str(),
-        //         "--path",
-        //         path.to_str().unwrap(),
-        //         "--broadcast",
-        //         format!("{}/{}/transaction/broadcast", self.endpoint, self.network).as_str(),
-        //         &name,
-        //     ])
-        //     .unwrap();
-        //
-        //     // Deploy program
-        //     Developer::Deploy(deploy).parse().map_err(CliError::failed_to_execute_deploy)?;
-        //
-        //     // Sleep for `wait_gap` seconds.
-        //     // This helps avoid parents from being serialized before children.
-        //     if index < all_paths.len() - 1 {
-        //         std::thread::sleep(std::time::Duration::from_secs(self.wait_gap));
-        //     }
-        // }
-
-        Err(PackageError::unimplemented_command("leo deploy").into())
+    fn apply(self, context: Context, _: Self::Input) -> Result<Self::Output> {
+        // Get the program name.
+        let project_name = context.open_manifest()?.program_id().to_string();
+
+        // Get the private key.
+        let mut private_key = self.fee_options.private_key;
+        if private_key.is_none() {
+            private_key =
+                Some(dotenv_private_key().map_err(CliError::failed_to_read_environment_private_key)?.to_string());
+        }
+
+        let mut all_paths: Vec<(String, PathBuf)> = Vec::new();
+
+        // Extract post-ordered list of local dependencies' paths from `leo.lock`.
+        if !self.non_recursive {
+            all_paths = context.local_dependency_paths()?;
+        }
+
+        // Add the parent program to be deployed last.
+        all_paths.push((project_name, context.dir()?.join("build")));
+
+        for (index, (name, path)) in all_paths.iter().enumerate() {
+            // Set the deploy arguments.
+            let mut deploy_args = vec![
+                "snarkos".to_string(),
+                "--private-key".to_string(),
+                private_key.as_ref().unwrap().clone(),
+                "--query".to_string(),
+                self.endpoint.clone(),
+                "--priority-fee".to_string(),
+                self.fee_options.priority_fee.to_string(),
+                "--path".to_string(),
+                path.to_str().unwrap().parse().unwrap(),
+                "--broadcast".to_string(),
+                format!("{}/{}/transaction/broadcast", self.endpoint, self.fee_options.network).to_string(),
+                name.clone(),
+            ];
+
+            // Use record as payment option if it is provided.
+            if let Some(record) = self.fee_options.record.clone() {
```

Comment:
If there are multiple programs, then the single record won't work. 
Would it make more sense to change the UX to individually deploy programs in the dependency tree?
The code could have additional checks to tell users what they need to deploy.

Reply:
That makes sense. What about changing it to defaulting to individually deploying programs, and allowing the option of recursive programs (and throwing an error if `--recursive` and `--record` are both set)?

Reply:
In terms of UX it is really nice being able to deploy a nested project all in CLI command